pyccel/parser/syntax/openmp.py

```diff
@@ -47,81 +49,178 @@ def expr(self):
+            combined = self.combined.expr
+        if isinstance(self.combined, OmpMaskedTaskloop):
+            _valid_clauses = _valid_clauses + (OmpFilter,)
+            if 'simd' in self.combined.expr:
+                _valid_clauses = _valid_clauses + _valid_simd_clauses
+            if 'taskloop' in self.combined.expr:
+                _valid_clauses = _valid_clauses + _valid_taskloop_clauses
+            combined = self.combined.expr
+        if isinstance(self.combined, OmpPSections):
+            _valid_clauses = _valid_clauses + _valid_sections_clauses
+            combined = self.combined.expr
         for clause in self.clauses:
             if isinstance(clause, _valid_clauses):
                 txt = '{0} {1}'.format(txt, clause.expr)
             else:
                 raise TypeError('Wrong clause for OmpParallelConstruct')
 
-        return OMP_Parallel_Construct(txt)
+        return OMP_Parallel_Construct(txt, combined)
 
 class OmpLoopConstruct(BasicStmt):
-    """Class representing a ."""
+    """Class representing a For loop construct."""
     def __init__(self, **kwargs):
         """
         """
         self.clauses = kwargs.pop('clauses')
 
-        super(OmpLoopConstruct, self).__init__(**kwargs)
+        super().__init__(**kwargs)
 
     @property
     def expr(self):
         if DEBUG:
             print("> OmpLoopConstruct: expr")
 
-        _valid_clauses = (OmpPrivate, \
-                         OmpFirstPrivate, \
-                         OmpLastPrivate, \
-                         OmpReduction, \
-                         OmpSchedule, \
-                         OmpCollapse, \
-                         OmpLinear, \
-                         OmpOrdered)
-
         txt = ''
         for clause in self.clauses:
-            if isinstance(clause, _valid_clauses):
+            if isinstance(clause, _valid_loop_clauses):
                 txt = '{0} {1}'.format(txt, clause.expr)
             else:
                 raise TypeError('Wrong clause for OmpLoopConstruct. Given : ', \
                                 type(clause))
         return OMP_For_Loop(txt)
 
+class OmpTaskLoopConstruct(BasicStmt):
+    """Class representing a taskloop construct."""
+    def __init__(self, **kwargs):
+        """
+        """
+        self.clauses = kwargs.pop('clauses')
+        self.name = kwargs.pop('name')
+
+        super().__init__(**kwargs)
+
+    @property
+    def expr(self):
+        if DEBUG:
+            print("> OmpTaskLoopConstruct: expr")
+
+        _valid_clauses = _valid_taskloop_clauses + (OmpinReduction,)
+
+        txt = self.name
+        for clause in self.clauses:
+            if isinstance(clause, _valid_clauses):
+                txt = '{0} {1}'.format(txt, clause.expr)
+            else:
+                raise TypeError('Wrong clause for OmpTaskLoopConstruct. Given : ', \
+                                type(clause))
+        return OmpAnnotatedComment(txt)
```

Comment:
See above, unnecessary calculations in `expr` is called multiple times

Reply:
im not sure i understand what do you mean by called multiple time

Reply:
I.e. If I do:
```
var = OmpTaskLoopConstruct(name=str1, clauses=str2)
tmp = var.expr
.....
f(var.expr)
```
Here I have accessed the property `var.expr` twice. So I have done the above calculation twice. If the calculation is in the `__init__` method then the calculation is only done once

Reply:
aah i see, i moved all the calculations to the `__init__` function.

Reply:
For example if you do:
```
def __init__(self, ...):
    ...
    self._expr = OmpAnnotatedComment(txt)
```
Then the property is simple:
```
@property
def expr(self):
    return self._expr
```

This is faster as there is less calculation, but it is also more memory efficient. With a `self._expr` we have `var.expr is var.expr == True` but with your current method `var.expr is var.expr == False`